From: Samuel Brian <samuel.brian@digi.com>
Date: Mon, 5 Aug 2019 14:15:27 +1000
Subject: bearer: reinstate "number" attribute

This reverts the parts of ModemManager commit
55c3026643ee6c75c0a68171abd884edf035b9ba that removed the "number" from the
Bearer Properties. More specifically, that commit left the API for the
"number" in place but made it non-functional.

The bearer "number" is used by patches for nominating PDP context ids for
bearers, which is required for dual APNs with MBIM and QMI muxing.
---
 cli/mmcli-bearer.c                |  3 +++
 libmm-glib/mm-bearer-properties.c | 38 ++++++++++++++++++-------------
 libmm-glib/mm-bearer-properties.h | 11 +++------
 3 files changed, 28 insertions(+), 24 deletions(-)

diff --git a/cli/mmcli-bearer.c b/cli/mmcli-bearer.c
index f8494626..4a475da7 100644
--- a/cli/mmcli-bearer.c
+++ b/cli/mmcli-bearer.c
@@ -160,6 +160,7 @@ print_bearer_info (MMBearer *bearer)
         const gchar *password = NULL;
         const gchar *rm_protocol = NULL;
         gchar       *allowed_auth_str = NULL;
+        const gchar *number = NULL;
 
         if (properties) {
             apn              = mm_bearer_properties_get_apn (properties);
@@ -169,6 +170,7 @@ print_bearer_info (MMBearer *bearer)
             password         = mm_bearer_properties_get_password (properties);
             if (mm_bearer_get_bearer_type (bearer) != MM_BEARER_TYPE_DEFAULT_ATTACH) {
                 roaming     = mm_bearer_properties_get_allow_roaming (properties) ? "allowed" : "forbidden";
+                number      = mm_bearer_properties_get_number (properties);
                 rm_protocol = mm_modem_cdma_rm_protocol_get_string (mm_bearer_properties_get_rm_protocol (properties));
             }
         }
@@ -180,6 +182,7 @@ print_bearer_info (MMBearer *bearer)
         mmcli_output_string           (MMC_F_BEARER_PROPERTIES_PASSWORD,     password);
         mmcli_output_string           (MMC_F_BEARER_PROPERTIES_RM_PROTOCOL,  rm_protocol);
         mmcli_output_string_list_take (MMC_F_BEARER_PROPERTIES_ALLOWED_AUTH, allowed_auth_str);
+        mmcli_output_string           (MMC_F_BEARER_PROPERTIES_NUMBER,       number);
     }
 
     /* IPv4 config */
diff --git a/libmm-glib/mm-bearer-properties.c b/libmm-glib/mm-bearer-properties.c
index 5ed4438e..0a3d37b6 100644
--- a/libmm-glib/mm-bearer-properties.c
+++ b/libmm-glib/mm-bearer-properties.c
@@ -38,12 +38,10 @@ G_DEFINE_TYPE (MMBearerProperties, mm_bearer_properties, G_TYPE_OBJECT);
 #define PROPERTY_USER            "user"
 #define PROPERTY_PASSWORD        "password"
 #define PROPERTY_IP_TYPE         "ip-type"
+#define PROPERTY_NUMBER          "number"
 #define PROPERTY_ALLOW_ROAMING   "allow-roaming"
 #define PROPERTY_RM_PROTOCOL     "rm-protocol"
 
-/* no longer used properties */
-#define DEPRECATED_PROPERTY_NUMBER "number"
-
 struct _MMBearerPropertiesPrivate {
     /* APN */
     gchar *apn;
@@ -51,6 +49,8 @@ struct _MMBearerPropertiesPrivate {
     MMBearerIpFamily ip_type;
     /* Allowed auth */
     MMBearerAllowedAuth allowed_auth;
+    /* Number */
+    gchar *number;
     /* User */
     gchar *user;
     /* Password */
@@ -303,8 +303,6 @@ mm_bearer_properties_get_allow_roaming (MMBearerProperties *self)
 
 /*****************************************************************************/
 
-#ifndef MM_DISABLE_DEPRECATED
-
 /**
  * mm_bearer_properties_set_number:
  * @self: a #MMBearerProperties.
@@ -322,7 +320,8 @@ mm_bearer_properties_set_number (MMBearerProperties *self,
 {
     g_return_if_fail (MM_IS_BEARER_PROPERTIES (self));
 
-    /* NO-OP */
+    g_free (self->priv->number);
+    self->priv->number = g_strdup (number);
 }
 
 /**
@@ -343,12 +342,9 @@ mm_bearer_properties_get_number (MMBearerProperties *self)
 {
     g_return_val_if_fail (MM_IS_BEARER_PROPERTIES (self), NULL);
 
-    /* NO-OP */
-    return NULL;
+    return self->priv->number;
 }
 
-#endif /* MM_DISABLE_DEPRECATED */
-
 /*****************************************************************************/
 
 /**
@@ -435,6 +431,12 @@ mm_bearer_properties_get_dictionary (MMBearerProperties *self)
                                PROPERTY_IP_TYPE,
                                g_variant_new_uint32 (self->priv->ip_type));
 
+    if (self->priv->number)
+        g_variant_builder_add (&builder,
+                               "{sv}",
+                               PROPERTY_NUMBER,
+                               g_variant_new_string (self->priv->number));
+
     if (self->priv->allow_roaming_set)
         g_variant_builder_add (&builder,
                                "{sv}",
@@ -499,7 +501,9 @@ mm_bearer_properties_consume_string (MMBearerProperties *self,
             return FALSE;
         }
         mm_bearer_properties_set_allow_roaming (self, allow_roaming);
-    } else if (g_str_equal (key, PROPERTY_RM_PROTOCOL)) {
+    } else if (g_str_equal (key, PROPERTY_NUMBER))
+        mm_bearer_properties_set_number (self, value);
+    else if (g_str_equal (key, PROPERTY_RM_PROTOCOL)) {
         GError *inner_error = NULL;
         MMModemCdmaRmProtocol protocol;
 
@@ -509,8 +513,6 @@ mm_bearer_properties_consume_string (MMBearerProperties *self,
             return FALSE;
         }
         mm_bearer_properties_set_rm_protocol (self, protocol);
-    } else if (g_str_equal (key, DEPRECATED_PROPERTY_NUMBER)) {
-        /* NO-OP */
     } else {
         g_set_error (error,
                      MM_CORE_ERROR,
@@ -598,13 +600,15 @@ mm_bearer_properties_consume_variant (MMBearerProperties *properties,
         mm_bearer_properties_set_ip_type (
             properties,
             g_variant_get_uint32 (value));
+    else if (g_str_equal (key, PROPERTY_NUMBER))
+        mm_bearer_properties_set_number (
+            properties,
+            g_variant_get_string (value, NULL));
     else if (g_str_equal (key, PROPERTY_ALLOW_ROAMING))
         mm_bearer_properties_set_allow_roaming (
             properties,
             g_variant_get_boolean (value));
-    else if (g_str_equal (key, DEPRECATED_PROPERTY_NUMBER)) {
-        /* NO-OP */
-    } else {
+    else {
         /* Set error */
         g_set_error (error,
                      MM_CORE_ERROR,
@@ -676,6 +680,7 @@ mm_bearer_properties_cmp (MMBearerProperties *a,
 {
     return ((!g_strcmp0 (a->priv->apn, b->priv->apn)) &&
             (a->priv->ip_type == b->priv->ip_type) &&
+            (!g_strcmp0 (a->priv->number, b->priv->number)) &&
             (a->priv->allowed_auth == b->priv->allowed_auth) &&
             (!g_strcmp0 (a->priv->user, b->priv->user)) &&
             (!g_strcmp0 (a->priv->password, b->priv->password)) &&
@@ -724,6 +729,7 @@ finalize (GObject *object)
     g_free (self->priv->apn);
     g_free (self->priv->user);
     g_free (self->priv->password);
+    g_free (self->priv->number);
 
     G_OBJECT_CLASS (mm_bearer_properties_parent_class)->finalize (object);
 }
diff --git a/libmm-glib/mm-bearer-properties.h b/libmm-glib/mm-bearer-properties.h
index 751185d9..0271a74d 100644
--- a/libmm-glib/mm-bearer-properties.h
+++ b/libmm-glib/mm-bearer-properties.h
@@ -70,6 +70,8 @@ void mm_bearer_properties_set_ip_type       (MMBearerProperties *self,
                                              MMBearerIpFamily ip_type);
 void mm_bearer_properties_set_allow_roaming (MMBearerProperties *self,
                                              gboolean allow_roaming);
+void mm_bearer_properties_set_number        (MMBearerProperties *self,
+                                             const gchar *number);
 void mm_bearer_properties_set_rm_protocol   (MMBearerProperties *self,
                                              MMModemCdmaRmProtocol protocol);
 
@@ -79,16 +81,9 @@ const gchar           *mm_bearer_properties_get_user          (MMBearerPropertie
 const gchar           *mm_bearer_properties_get_password      (MMBearerProperties *self);
 MMBearerIpFamily       mm_bearer_properties_get_ip_type       (MMBearerProperties *self);
 gboolean               mm_bearer_properties_get_allow_roaming (MMBearerProperties *self);
+const gchar           *mm_bearer_properties_get_number        (MMBearerProperties *self);
 MMModemCdmaRmProtocol  mm_bearer_properties_get_rm_protocol   (MMBearerProperties *self);
 
-#ifndef MM_DISABLE_DEPRECATED
-G_DEPRECATED
-void         mm_bearer_properties_set_number (MMBearerProperties *self,
-                                              const gchar        *number);
-G_DEPRECATED
-const gchar *mm_bearer_properties_get_number (MMBearerProperties *self);
-#endif
-
 /*****************************************************************************/
 /* ModemManager/libmm-glib/mmcli specific methods */
 
